src/Java.Interop.Tools.JavaCallableWrappers/Java.Interop.Tools.JavaCallableWrappers/JavaTypeScanner.cs

```diff
@@ -29,23 +29,38 @@ public List<TypeDefinition> GetJavaTypes (IEnumerable<string> assemblies, IAssem
 
 			foreach (var assembly in assemblies) {
 				var assm = resolver.GetAssembly (assembly);
-
 				foreach (ModuleDefinition md in assm.Modules) {
 					foreach (TypeDefinition td in md.Types) {
-						AddJavaTypes (javaTypes, td);
+						AddJavaTypes (javaTypes.Add, td);
 					}
 				}
 			}
 
 			return javaTypes;
 		}
 
-		void AddJavaTypes (List<TypeDefinition> javaTypes, TypeDefinition type)
+		public List<TypeDefinition> GetJavaTypesInParallel (IEnumerable<string> assemblies, IAssemblyResolver resolver)
+		{
+			var javaTypes = new BlockingCollection<TypeDefinition> ();
+
+			Parallel.ForEach (assemblies, assembly => {
+				var assm = resolver.GetAssembly (assembly);
```

Comment:
Nope. Nopity nopity not going to do it ***NOPE***.

Cecil isn't thread safe.  This ***cannot*** work.  See also: b1667a2be6ce12c5057abc38de8f9ee35d362a65.

Reply:
Hmm b1667a2, looks like it is parallelizing on the `TypeDefinition` and this is doing a `Task` per `AssemblyDefinition`.

Let me try writing a test like you have here, and see what happens.